### PR TITLE
added solution method

### PR DESCRIPTION
**Pull request by tsehelnyk**

### CODE REVIEW

File: src/main/java/core/basesyntax/ArrayList.java [hunk 1]
```diff
@@ -1,53 +1,126 @@
...
+            throw new IllegalArgumentException("Illegal Capacity: " + capacity);
+        }
+    }
+
+    public ArrayList() {
+        elementData = (T[]) new Object[DEFAULT_CAPACITY];
+    }
+
+    private void grow() {
+        T[] temp = (T[]) new Object[Math.round(elementData.length * 3 / 2) + 1];
```
Comment (Tarabarnik) [inline src/main/java/core/basesyntax/ArrayList.java]: try to solve without Math library

File: src/main/java/core/basesyntax/ArrayList.java [hunk 2]
```diff
@@ -1,53 +1,126 @@
...
+        temp = Arrays.copyOf(elementData, temp.length);
+        elementData = temp;
+    }
+
+    private boolean validIndex(int index) {
+        if (index < 0 || !(index < size)) {
```
Comment (Tarabarnik) [inline src/main/java/core/basesyntax/ArrayList.java]:
```suggestion
        if (index < 0 || index >= size) {
```

File: src/main/java/core/basesyntax/ArrayList.java [hunk 3]
```diff
@@ -1,53 +1,126 @@
...
+        elementData = (T[]) new Object[DEFAULT_CAPACITY];
+    }
+
+    private void grow() {
+        T[] temp = (T[]) new Object[Math.round(elementData.length * 3 / 2) + 1];
+        temp = Arrays.copyOf(elementData, temp.length);
+        elementData = temp;
+    }
+
+    private boolean validIndex(int index) {
```
Comment (Tarabarnik) [inline src/main/java/core/basesyntax/ArrayList.java]: make it void, please

File: src/main/java/core/basesyntax/ArrayList.java [hunk 4]
```diff
@@ -1,53 +1,126 @@
...
+        if (index < 0 || !(index < size)) {
+            throw new ArrayIndexOutOfBoundsException("Illegal Index: " + index);
+        }
+        return true;
+    }
+
     @Override
     public void add(T value) {
-
+        if (elementData.length > size + 1) {
```
Comment (Tarabarnik) [inline src/main/java/core/basesyntax/ArrayList.java]:
```suggestion
        if (elementData.length >= size) {
```

File: src/main/java/core/basesyntax/ArrayList.java [hunk 5]
```diff
@@ -1,53 +1,126 @@
...
+            elementData[size] = value;
+            size++;
+        } else {
```
Comment (Tarabarnik) [inline src/main/java/core/basesyntax/ArrayList.java]: try to solve without else block

File: src/main/java/core/basesyntax/ArrayList.java [hunk 6]
```diff
@@ -1,53 +1,126 @@
...
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (this.validIndex(index) && elementData.length > size + 1) {
+            System.arraycopy(elementData, index, elementData, index + 1, size - index);
+            elementData[index] = value;
+            size++;
+        } else {
```
Comment (Tarabarnik) [inline src/main/java/core/basesyntax/ArrayList.java]: same

File: src/main/java/core/basesyntax/ArrayList.java [hunk 7]
```diff
@@ -1,53 +1,126 @@
...
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        if ((elementData.length - size) > (list.size() * 2)) {
+            for (int i = 0; i < list.size(); i++) {
+                this.add(list.get(i));
+            }
+        } else {
```
Comment (Tarabarnik) [inline src/main/java/core/basesyntax/ArrayList.java]: same

File: src/main/java/core/basesyntax/ArrayList.java (hunk 1 above)
Comment (tsehelnyk) [reply src/main/java/core/basesyntax/ArrayList.java]: ok

File: src/main/java/core/basesyntax/ArrayList.java (hunk 2 above)
Comment (tsehelnyk) [reply src/main/java/core/basesyntax/ArrayList.java]: ok

File: src/main/java/core/basesyntax/ArrayList.java (hunk 4 above)
Comment (tsehelnyk) [reply src/main/java/core/basesyntax/ArrayList.java]: ok

File: src/main/java/core/basesyntax/ArrayList.java (hunk 3 above)
Comment (tsehelnyk) [reply src/main/java/core/basesyntax/ArrayList.java]: done

File: src/main/java/core/basesyntax/ArrayList.java (hunk 5 above)
Comment (tsehelnyk) [reply src/main/java/core/basesyntax/ArrayList.java]: ok

File: src/main/java/core/basesyntax/ArrayList.java (hunk 6 above)
Comment (tsehelnyk) [reply src/main/java/core/basesyntax/ArrayList.java]: ok

File: src/main/java/core/basesyntax/ArrayList.java (hunk 7 above)
Comment (tsehelnyk) [reply src/main/java/core/basesyntax/ArrayList.java]: ok

File: src/main/java/core/basesyntax/ArrayList.java [hunk 8]
```diff
@@ -1,53 +1,115 @@
...
+    private void validIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new ArrayIndexOutOfBoundsException("Illegal Index: " + index);
+        }
+    }
+
     @Override
     public void add(T value) {
-
+        if (!(elementData.length > size)) {
```
Comment (Tarabarnik) [inline src/main/java/core/basesyntax/ArrayList.java]:
```suggestion
        if (elementData.length <= size) {
```

File: src/main/java/core/basesyntax/ArrayList.java [hunk 9]
```diff
@@ -1,53 +1,115 @@
...
+        }
+        elementData[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        this.validIndex(index);
+        if (!(elementData.length > size)) {
```
Comment (Tarabarnik) [inline src/main/java/core/basesyntax/ArrayList.java]: same

File: src/main/java/core/basesyntax/ArrayList.java [hunk 10]
```diff
@@ -1,53 +1,115 @@
...
+        }
+        System.arraycopy(elementData, index, elementData, index + 1, size - index);
+        elementData[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        if (!((elementData.length - size) > (list.size() * 2))) {
```
Comment (Tarabarnik) [inline src/main/java/core/basesyntax/ArrayList.java]: this check is redundant

File: src/main/java/core/basesyntax/ArrayList.java [hunk 11]
```diff
@@ -1,53 +1,115 @@
...
+        this.validIndex(index);
+        T removed = elementData[index];
+        System.arraycopy(elementData, index + 1, elementData, index, size-- - index);
+        elementData[size] = null;
+        return removed;
     }
 
     @Override
     public T remove(T t) {
+        if (size > 0) {
```
Comment (Tarabarnik) [inline src/main/java/core/basesyntax/ArrayList.java]: try to solve without this check

**Pull request closed without merging**